vulnerabilities/importers/ruby.py

```diff
@@ -1,76 +1,126 @@
+                recursive=True, file_ext='yml', subdir='./gems')
+
+    def updated_advisories(self) -> Set[Advisory]:
+        files = self._updated_files.union(self._added_files)
+        advisories = []
+        for f in files:
+            processed_data = self.process_file(f)
+            if processed_data:
+                advisories.append(processed_data)
+        return self.batch_advisories(advisories)
+
+    def process_file(self, path) -> List[Advisory]:
+        with open(path) as f:
+            record = yaml.safe_load(f)
+            package_name = record.get(
+                'gem')
+
+            if not package_name:
+                return
+
+            if 'cve' in record:
+                cve_id = 'CVE-{}'.format(record['cve'])
+            else:
+                return
+
+            safe_version_ranges = record.get('patched_versions', [])
+            # this case happens when the advisory contain only 'patched_versions' field
+            # and it has value None(i.e it is empty :( ).
+            if not safe_version_ranges:
+                safe_version_ranges = []
+            safe_version_ranges += record.get('unaffected_versions', [])
+            safe_version_ranges = [i for i in safe_version_ranges if i]
+
+            if not getattr(self, 'pkg_manager_api', None):
+                self.pkg_manager_api = rubyAPI()
+            all_vers = self.pkg_manager_api.get_all_version_of_package(
+                package_name)
+            safe_versions, affected_versions = self.categorize_versions(
+                all_vers, safe_version_ranges)
+
+            impacted_purls = {
+                PackageURL(
+                    name=package_name,
+                    type='gem',
+                    version=version,
+                ) for version in affected_versions}
+
+            resolved_purls = {
+                PackageURL(
+                    name=package_name,
+                    type='gem',
+                    version=version,
+                ) for version in safe_versions}
+
+            return Advisory(
+                summary=record.get('description', ''),
+                impacted_package_urls=impacted_purls,
+                resolved_package_urls=resolved_purls,
+                reference_urls=record.get('url', ''),
+                cve_id=cve_id
+            )
+
+    @staticmethod
+    def categorize_versions(all_versions, unaffected_version_ranges):
+
+        for id, elem in enumerate(unaffected_version_ranges):
+            try:
+                unaffected_version_ranges[id] = RangeSpecifier(
+                    elem.replace(' ', ''))
+            except InvalidSpecifier:
+                continue
+
+        safe_versions = set()
+        for i in all_versions:
+            for ver_rng in unaffected_version_ranges:
+
+                if i in ver_rng:
+
+                    safe_versions.add(i)
+
+        return (safe_versions, all_versions-safe_versions)
+
+
+class rubyAPI:
```

Comment:
The reason I added these `VersionAPI` classes in other importers was the caching. Apart from it being useful in tests to pass an instance with a pre-seeded cache, they avoid making multiple requests for the same package. Since the lifetime of these objects is only the duration of one import run, I don't think we will have issues with stale data in the cache.

However, using `requests.Session` also has benefits, like connection pooling. So we probably want both in all of these classes. I created a ticket for updating the other ones: #187. Do you think it makes sense to add caching to this class?

Reply:
We eventually want to club these `VersionAPI` into a single module, that will also help us with implementing an importer for https://github.com/nexB/vulnerablecode/issues/74 since we would need to query multiple package managers.  

There should  be a way to do the tests without caching. But your point of avoiding querying for the same package multiple times is strong and I believe it won't hurt, I'll implement cache.

Reply:
> There should be a way to do the tests without caching. But your point of avoiding querying for the same package multiple times is strong and I believe it won't hurt, I'll implement cache.

What problem do you see with using the cache in tests? I think that approach, or the one you took here with `patch` and defining a return value both work fine. We just want predictable data so we can make precise assertions and avoid network I/O happening in the tests. Both approaches achieve that, right?

Reply:
Yes. I am adding cache to the importer.